### PR TITLE
[GH-971] Support Jira migration to Oauth

### LINKED ISSUE #971: Support migration of existing Jira Cloud connection to use OAuth approach
In order to migrate an existing MM instance currently using Jira Cloud to use the OAuth approach, we need to add a way to simultaneously support both instance types from the same Jira URL. This is so webhooks/subscriptions etc. can continue to function properly, and existing connected users can continue to interact with the plugin, as the user base is in the process of migrating connections to OAuth.

_Issue created from a [Mattermost message](https://community.mattermost.com/_redirect/pl/qpjkoimuhfn4frqf9ky5eticsw) *by @mickmister*._

### PR DESCRIPTION
**Pull request by raghavaggarwal2308**
#### Summary
- While installing the oauth instance we are storing the previously stored JWT instance data inside the new oauth instance data.
- While getting a client for making the API requests we are returning a JWT client in case the stored JWT instance is not nil and the user's oauth token is nil.
- We have added several debug statements to keep track of how the code is working as there was no way for us to test this change properly. (Since we don't have a server with a JWT (cloud instance) setup)

#### What to test?
- If an instance was previously connected to Jira using JWT (cloud instance) and the admin has installed a cloud-oauth instance then all the users who have not connected to oauth yet should be able to continue using the plugin.

#### Steps to reproduce:
- Install a cloud-oauth instance on a server that already has a JWT (cloud instance) setup and certain users have connected to it.

#### Ticket Link

Fixes #971

### CODE REVIEW

File: server/instances.go [hunk 1]
```diff
@@ -151,6 +151,29 @@ func (p *Plugin) InstallInstance(instance Instance) error {
...
+					p.API.LogError("Error occurred while fetching the instance", "ID", instance.GetID(), "Error", err.Error())
+					return err
+				}
+
+				// Storing the JWT instance data inside the OAuth instance data. We will use this to use the stored JWT token in case the user has not connected to OAuth yet.
+				p.API.LogDebug("Instance type stored in KV store", "ID", jwtInstance.GetID(), "Type", jwtInstance.Common().Type)
+				if jwtInstance.Common().Type == CloudInstanceType {
+					instance.(*cloudOAuthInstance).JWTInstance = jwtInstance.(*cloudInstance)
+				} else if jwtInstance.Common().Type == CloudOAuthInstanceType {
+					instance.(*cloudOAuthInstance).JWTInstance = jwtInstance.(*cloudOAuthInstance).JWTInstance
```
Comment (mickmister) [inline server/instances.go:167]: This seems a bit weird to me but also makes sense. Is the point of this line is to make sure we handle the case where we run `/jira instance install cloud-oauth` twice? We "carry over" the JWT instance embedded inside the existing one. This seems a bit error-prone since this is the only field we're carrying over. Are there any other fields we want to retain when "over-installing" a `cloud-oauth` over an existing `cloud-oauth` instance?

Maybe put a comment here stating this is here to make sure we carry it over when we install `cloud-oauth` twice

File: server/instance_cloud_oauth.go [hunk 2]
```diff
@@ -99,6 +99,11 @@ func (ci *cloudOAuthInstance) getClientForConnection(connection *Connection) (*j
 	oauth2Conf := ci.GetOAuthConfig()
 	ctx := context.Background()
 	tokenSource := oauth2Conf.TokenSource(ctx, connection.OAuth2Token)
+	if ci.JWTInstance != nil && tokenSource == nil {
```
Comment (mickmister) [inline server/instance_cloud_oauth.go:102]: What are the benefits of using `tokenSource == nil` here instead of `connection.OAuth2Token == nil`?

File: server/instance_cloud_oauth.go [hunk 3]
```diff
@@ -99,6 +99,11 @@ func (ci *cloudOAuthInstance) getClientForConnection(connection *Connection) (*j
...
+		ci.Plugin.API.LogDebug("Returning a JWT token client in case the stored JWT instance is not nil and the user's oauth token is nil")
```
Comment (mickmister) [inline server/instance_cloud_oauth.go:103]:
```suggestion
		ci.Plugin.API.LogDebug("Returning a JWT token client since the stored JWT instance is not nil and the user's oauth token is nil")
```

File: server/instances.go [hunk 4]
```diff
@@ -151,6 +151,29 @@ func (p *Plugin) InstallInstance(instance Instance) error {
 				}
 			}
 
+			if instance.Common().Type == CloudOAuthInstanceType && len(instances.IDs()) > 0 && instances.checkIfExists(instance.GetID()) {
```
Comment (mickmister) [inline server/instances.go:154]: I originally thought we also want to check if the previous instance is `cloud` in this if condition, but I see in a block below we have a special case for when the previous instance was `cloud-oauth` as well

File: server/instances.go (hunk 1 above)
Comment (mickmister) [reply server/instances.go:167]: It seems like a null pointer is not possible here but still a little concerned about it. Ideally we always do the `ok` syntax to avoid type assertion panics

File: server/instances.go (hunk 4 above)
Comment (mickmister) [inline server/instances.go:154]: Let's do the type assertion once at the top of this function, to avoid the multiple type assertions below

```suggestion
			if instance.Common().Type == CloudOAuthInstanceType && len(instances.IDs()) > 0 && instances.checkIfExists(instance.GetID()) {
			    cloudOAuthInstance, ok := instance.(*cloudOAuthInstance)
			    if !ok {
			        p.API.LogError("Instance type is `cloud-oauth` but failed to assert instance.(*cloudOAuthInstance).", "ID", instance.GetID())
			    } else {
			        // ...all other logic
```

File: server/instances.go [hunk 5]
```diff
@@ -151,6 +151,29 @@ func (p *Plugin) InstallInstance(instance Instance) error {
...
+				p.API.LogDebug("Getting the stored JWT instance data to store it inside the OAuth instance data.")
+				jwtInstance, err := p.instanceStore.LoadInstance(instance.GetID())
```
Comment (mickmister) [inline server/instances.go:156]:
```suggestion
				previousInstance, err := p.instanceStore.LoadInstance(instance.GetID())
```

We should probably also rename the function parameter `instance` to `newInstance` as well

File: server/instances.go [hunk 6]
```diff
@@ -151,6 +151,29 @@ func (p *Plugin) InstallInstance(instance Instance) error {
...
+				if jwtInstance.Common().Type == CloudInstanceType {
+					instance.(*cloudOAuthInstance).JWTInstance = jwtInstance.(*cloudInstance)
+				} else if jwtInstance.Common().Type == CloudOAuthInstanceType {
+					instance.(*cloudOAuthInstance).JWTInstance = jwtInstance.(*cloudOAuthInstance).JWTInstance
+				}
+
+				if instance.(*cloudOAuthInstance).JWTInstance != nil {
+					p.API.LogDebug("JWT instance is successfully stored inside cloud OAuth instance data.", "JWTInstance", instance.(*cloudOAuthInstance).JWTInstance)
+				} else {
+					p.API.LogDebug("Failed to store JWT instance inside cloud OAuth instance data or there was no JWT instance installed previously.", "JWTInstance", instance.(*cloudOAuthInstance).JWTInstance)
```
Comment (mickmister) [inline server/instances.go:173]: The "Failed to store JWT instance" piece of this seems alarming, especially since this is typically not an "error". Maybe something like this instead:

```suggestion
					p.API.LogDebug("No JWT instance inside cloud OAuth instance data.")
```

**Review by mickmister — COMMENTED** [review]
Awesome work on this @raghavaggarwal2308 :+1:

In general LGTM, just some comments about variable names and avoiding panics

---

Another thing that comes to mind is the webhook events for subscriptions, particularly comment events. Does this PR make it so when there is a comment event and the instance is `cloud-oauth`, do we:

- if comment author is connected, use their token to fetch the comment
- else if `JWTInstance != nil` use the "bot" to fetch the comment

File: server/instances.go (hunk 1 above)
Comment (raghavaggarwal2308) [reply server/instances.go:167]: @mickmister This function "installInstance" is being called from the "installCloudOAuthInstance" function. After running the install command "installCloudOAuthInstance" is called twice during the flow:
1. When we run the install command.
2. When we submit the clientID and clientSecret modal. (To store the credentials in the instance data)

The instance is stored in the KV store after running the install command and the JWT instance data is stored inside it (Now the previous JWT instance data is overridden by the oauth instance data). When the function is called on submitting the modal, if we don't carry forward the JWT instance data then it will be lost.

**Comment by raghavaggarwal2308** [comment]
>Another thing that comes to mind is the webhook events for subscriptions, particularly comment events. Does this PR make it so when there is a comment event and the instance is cloud-oauth, do we:
> - if the comment author is connected, use their token to fetch the comment
> - else if JWTInstance != nil use the "bot" to fetch the comment

@mickmister
- If the user is connected to oauth it will use the token to fetch the comment.
- If the user is not connected to oauth it will use the JWT token to fetch the comment.

In both cases the comment author needs to be connected. Do you want us to modify the logic to use the "bot" to fetch the comment in the second case?

File: server/instance_cloud_oauth.go (hunk 2 above)
Comment (raghavaggarwal2308) [reply server/instance_cloud_oauth.go:102]: @mickmister There are no particular benefits here but we are using `connection.OAuth2Token` now to avoid some extra computations of refreshing the token.

**Comment by mickmister** [comment]
> In both cases the comment author needs to be connected. Do you want us to modify the logic to use the "bot" to fetch the comment in the second case?

@raghavaggarwal2308 I'm thinking let's support a third case of "user is not connected to `cloud-oauth` or `cloud`, so fall back to using bot"

File: server/instances.go (hunk 1 above)
Comment (mickmister) [reply server/instances.go]: > When the function is called on submitting the modal, if we don't carry forward the JWT instance data then it will be lost.

Yeah when reviewing the OAuth implementation I thought it was weird how we call "install" when submitting the credential form. Being able to run `/jira install` on an existing instance was also introduced as a result, which also seemed off.

The fact that we have to manually add the `JWTInstance` here without making sure we copy everything over is a little concerning. If we add more fields to the `cloud-oauth` instance, we have to update this piece of code, and that seems error-prone.

Also I wonder if there are unwanted side effects to running `/jira install server myurl` after an existing Jira Server has been installed with that same URL. We may potentially lose information on the existing struct.

---

I think we can make the code more defensive by changing the "clientSecret modal" logic to be:

```go
if existingInstance != nil {
    existingInstance.ClientID = clientID
    existingInstance.ClientSecret = clientSecret
    storeInstance(existingInstance)
} else {
    newInstance := makeNewInstance()
    installInstance(newInstance)
}
```

And then disallow calling `installInstance` with an existing instance, just as it was before the OAuth code was introduced. `/jira install` can just start a setup flow if an instance exists.

File: server/instance_cloud_oauth.go [hunk 7]
```diff
@@ -98,12 +98,12 @@ func (ci *cloudOAuthInstance) GetClient(connection *Connection) (Client, error)
 func (ci *cloudOAuthInstance) getClientForConnection(connection *Connection) (*jira.Client, *http.Client, error) {
 	oauth2Conf := ci.GetOAuthConfig()
 	ctx := context.Background()
-	tokenSource := oauth2Conf.TokenSource(ctx, connection.OAuth2Token)
-	if ci.JWTInstance != nil && tokenSource == nil {
-		ci.Plugin.API.LogDebug("Returning a JWT token client in case the stored JWT instance is not nil and the user's oauth token is nil")
+	if ci.JWTInstance != nil && connection.OAuth2Token == nil {
+		ci.Plugin.API.LogDebug("Returning a JWT token client since the stored JWT instance is not nil and the user's oauth token is nil")
```
Comment (mickmister) [inline server/instance_cloud_oauth.go:102]: If I understand correctly, the `connection` variable here is ambiguous whether the user's connection is `cloud` or `cloud-oauth`. Maybe put a comment here that says "Checking if this user's connection is for a JWT instance"

Something else comes to mind. If the user disconnects their account, we should make sure we do proper cleanup of anything related to JWT instance type, in case that was their connection

File: server/instances.go [hunk 8]
```diff
@@ -141,44 +141,49 @@ func (p *instancesArray) Resize(n int) {
...
+						p.API.LogError("Error occurred while fetching the instance", "ID", newInstance.GetID(), "Error", err.Error())
+						return err
+					}
+
+					// Storing the JWT instance data inside the OAuth instance data. We will use this to use the stored JWT token in case the user has not connected to OAuth yet.
+					p.API.LogDebug("Instance type stored in KV store", "ID", previousInstance.GetID(), "Type", previousInstance.Common().Type)
+					if previousInstance.Common().Type == CloudInstanceType {
+						oAuthInstance.JWTInstance = previousInstance.(*cloudInstance)
+					} else if previousInstance.Common().Type == CloudOAuthInstanceType {
+						oAuthInstance.JWTInstance = previousInstance.(*cloudOAuthInstance).JWTInstance
```
Comment (mickmister) [inline server/instances.go:171]: We're doing some more type assertions here. I think we should be more defensive and use the `ok` syntax here

File: server/instance_cloud_oauth.go (hunk 7 above)
Comment (raghavaggarwal2308) [reply server/instance_cloud_oauth.go:102]: > Something else comes to mind. If the user disconnects their account, we should make sure we do proper cleanup of anything related to JWT instance type, in case that was their connection

@mickmister We are deleting the user's connection when a user is disconnecting. So, I think that should cover what you are suggesting here.
We cannot wipe the JWT instance data completely because even if a user has disconnected their account, there might be other users who are still using JWT.

**Comment by raghavaggarwal2308** [comment]
> @raghavaggarwal2308 I'm thinking let's support a third case of "user is not connected to cloud-oauth or cloud, so fall back to using bot". That bot thing could stop working at any point, so it should be the last resort

@mickmister I have added the fallback. Just to avoid any confusion, the fallback will only work if there is stored JWT instance as the bot uses JWT for authentication.

**Comment by hanzei** [comment]
@mickmister I don't have much knowledge about the token handling in the JIRA plugin. Would you mind asking another person fro a review?

File: server/instance_cloud_oauth.go (hunk 7 above)
Comment (mickmister) [reply server/instance_cloud_oauth.go]: > We cannot wipe the JWT instance data completely because even if a user has disconnected their account, there might be other users who are still using JWT.

Right, I worded "we should make sure we do proper cleanup of anything related to JWT instance type" a bit misleadingly. I really meant cleaning up the connection itself, and not the centralized JWT instance data. LGTM :+1:

File: server/instance_cloud_oauth.go [hunk 9]
```diff
@@ -98,12 +98,14 @@ func (ci *cloudOAuthInstance) GetClient(connection *Connection) (Client, error)
 func (ci *cloudOAuthInstance) getClientForConnection(connection *Connection) (*jira.Client, *http.Client, error) {
 	oauth2Conf := ci.GetOAuthConfig()
 	ctx := context.Background()
-	tokenSource := oauth2Conf.TokenSource(ctx, connection.OAuth2Token)
-	if ci.JWTInstance != nil && tokenSource == nil {
-		ci.Plugin.API.LogDebug("Returning a JWT token client in case the stored JWT instance is not nil and the user's oauth token is nil")
+
+	// Checking if this user's connection is for a JWT instance
+	if ci.JWTInstance != nil && connection.OAuth2Token == nil {
+		ci.Plugin.API.LogDebug("Returning a JWT token client since the stored JWT instance is not nil and the user's oauth token is nil")
```
Comment (mickmister) [inline server/instance_cloud_oauth.go:104]: Are you thinking we should remove log statements like this in a following release?

File: server/instances.go (hunk 1 above)
Comment (mickmister) [reply server/instances.go]: @raghavaggarwal2308 What do you think about my comment above?

File: server/webhook_jira.go [hunk 10]
```diff
@@ -52,9 +52,20 @@ func (jwh *JiraWebhook) expandIssue(p *Plugin, instanceID types.ID) error {
 			}
 
 			jwh.Issue = *issue
-		} else if _, ok := instance.(*cloudOAuthInstance); ok {
+		} else if instance, ok := instance.(*cloudOAuthInstance); ok {
 			mmUserID, err := p.userStore.LoadMattermostUserID(instanceID, jwh.Comment.Author.AccountID)
 			if err != nil {
+				if instance.JWTInstance != nil {
```
Comment (mickmister) [inline server/webhook_jira.go:58]: I think this would be easier to read if we returned early here with `instance.JWTInstance == nil ` instead of `instance.JWTInstance != nil `. Also added comments and an error message

```go
if err != nil {
    // User is not connected, so we try to fall back to JWT bot
    if instance.JWTInstance == nil {
        return errors.Wrap(err, "Cannot create subscription posts for this comment as the Jira comment author is not connected to Mattermost.")
    }

    // Fetch issue details with bot JWT bot
    var issue *jira.Issue
    issue, err = p.getIssueDataForCloudWebhook(instance.JWTInstance, jwh.Issue.ID)
    if err != nil {
        return errors.Wrap(err, "failed to getIssueDataForCloudWebhook using bot account")
    }

    jwh.Issue = *issue
    return nil
}
```

File: server/instance_cloud_oauth.go (hunk 9 above)
Comment (raghavaggarwal2308) [reply server/instance_cloud_oauth.go:104]: @mickmister Yes, we will remove these in a later release. These are just to check if the code executed as we expected it to.

File: server/instances.go (hunk 1 above)
Comment (raghavaggarwal2308) [reply server/instances.go]: @mickmister I agree with your suggestion. I have made the necessary changes according to the same.

Note: I have updated the condition to override the enterprise check to work only when a user is migrating from JWT to OAuth. Now if we try to run the "/jira install cloud-oauth ..." OR "/jira install server ..." command twice then we will get that message "You need a valid enterprise licence ....".

**Comment by raghavaggarwal2308** [comment]
@mickmister Fixed the review comments mentioned by you

**Comment by mickmister** [comment]
Hi @raghavaggarwal2308 I was going to request some changes, namely moving the backwards compatible `oauthInstance.JWTInstance` assignment from `p.InstallInstance` to `p.installCloudOAuthInstance`, though I just implemented them myself in the interest of time

For some reason, GitHub is not allowing me to add commits to your branch. Can you please merge the changes from https://github.com/mattermost/mattermost-plugin-jira/tree/MM-971-refactor into your branch? Please leave any comments you have on those changes on this PR. Thank you

File: server/instance_cloud_oauth.go [hunk 11]
```diff
@@ -71,20 +72,53 @@ func (p *Plugin) installCloudOAuthInstance(rawURL, clientID, clientSecret string
...
+
+			newInstance.JWTInstance = oauthInstance.JWTInstance
+			if newInstance.JWTInstance != nil {
+				p.API.LogDebug("Installing cloud-oauth over existing cloud-oauth instance. Carrying over existing saved JWT instance.")
+			} else {
+				p.API.LogDebug("Installing cloud-oauth over existing cloud-oauth instance. There exists no previous JWT instance to carry over.")
+			}
+		}
+
+		if existingInstance.Common().Type == CloudInstanceType {
```
Comment (raghavaggarwal2308) [inline server/instance_cloud_oauth.go:104]: Do we need to make it an `else if` condition to avoid an extra check?

File: server/instances.go [hunk 12]
```diff
@@ -145,53 +145,19 @@ func (p *Plugin) InstallInstance(newInstance Instance) error {
...
 			}
 
-			if !p.enterpriseChecker.HasEnterpriseFeatures() {
-				if instances != nil && len(instances.IDs()) > 0 {
-					// Override the enterprise check if a user is migrating from JWT to OAuth
-					if !(instances.checkIfExists(newInstance.GetID()) && previousInstance.Common().Type == CloudInstanceType && newInstance.Common().Type == CloudOAuthInstanceType) {
-						return errors.Errorf(licenseErrorString)
-					}
-				}
+			if !p.enterpriseChecker.HasEnterpriseFeatures() && len(instances.IDs()) > 0 && !instances.checkIfExists(newInstance.GetID()) {
```
Comment (raghavaggarwal2308) [inline server/instances.go:152]: We are allowing to install multiple instances over each other with the same instance ID. Do we need to support this?

**Comment by raghavaggarwal2308** [comment]
@mickmister Added a few minor comments, otherwise LGTM.

File: server/instances.go (hunk 12 above)
Comment (mickmister) [reply server/instances.go:152]: @raghavaggarwal2308 I decided to go with this to minimize risk on the changes here, since this is how it was already working before this PR. I went through the install flow a few times with `/jira setup` and `/jira instance install cloud-oauth` `/jira instance uninstall cloud-oaith` several times in different orders to try and generate some issue, and it seemed to all work flawlessly in every case.

I think having it how it currently works is fine. Installing over the same instance effectively just kicks off the setup flow for that instance type.

File: server/instance_cloud_oauth.go [hunk 13]
```diff
@@ -70,21 +72,53 @@ func (p *Plugin) installCloudOAuthInstance(rawURL, clientID, clientSecret string
...
```
Comment (mickmister) [inline server/instance_cloud_oauth.go:104]:
```suggestion
		} else if existingInstance.Common().Type == CloudInstanceType {
```

File: server/instance_cloud_oauth.go (hunk 11 above)
Comment (mickmister) [reply server/instance_cloud_oauth.go:104]: I used regular ifs so that there could be whitespace between the blocks to separate the two cases, though I think it reads better as you suggest. Please take a look at my suggested change below and add to the PR if GH allows you to do so through the UI

**Review by mickmister — APPROVED** [review]

File: server/setup_flow.go [hunk 14]
```diff
@@ -633,9 +633,20 @@ func (p *Plugin) submitCreateCloudOAuthInstance(f *flow.Flow, submission map[str
...
-	jiraURL, instance, err := p.installCloudOAuthInstance(jiraURL, clientID, clientSecret)
+	existingInstance, err := p.instanceStore.LoadInstance(types.ID(jiraURL))
 	if err != nil {
-		return "", nil, nil, err
+		return "", nil, nil, errors.Wrap(err, "failed to load existing cloud-oauth instance")
+	}
+
+	instance, ok := existingInstance.(*cloudOAuthInstance)
+	if !ok {
+		return "", nil, nil, errors.Errorf("existing instance is not a cloud-oauth instance. ID: %s", jiraURL)
```
Comment (esarafianou) [inline server/setup_flow.go:643]: Why does the `submitCreateCloudOAuthInstance` assume that there's already an existing instance of cloud-oauth? Isn't the `submitCreateCloudOAuthInstance` what get's triggered once the admin has clicked the submit button? That should work regardless of whether there is an existing instance.

File: server/setup_flow.go [hunk 15]
```diff
@@ -633,9 +633,20 @@ func (p *Plugin) submitCreateCloudOAuthInstance(f *flow.Flow, submission map[str
...
+	}
+
+	instance.JiraClientID = clientID
+	instance.JiraClientSecret = clientSecret
+	if err = p.instanceStore.StoreInstance(instance); err != nil {
```
Comment (esarafianou) [inline server/setup_flow.go:648]: Why are we storing the clientID and clientSecret of the new oauth instance to the existing one?

File: server/instance_cloud_oauth.go [hunk 16]
```diff
@@ -70,21 +72,51 @@ func (p *Plugin) installCloudOAuthInstance(rawURL, clientID, clientSecret string
 		return "", nil, err
 	}
 
-	instance := &cloudOAuthInstance{
-		InstanceCommon:   newInstanceCommon(p, CloudOAuthInstanceType, types.ID(jiraURL)),
-		MattermostKey:    p.GetPluginKey(),
-		JiraClientID:     clientID,
-		JiraClientSecret: clientSecret,
```
Comment (esarafianou) [inline server/instance_cloud_oauth.go:77]: Why are these removed?

File: server/setup_flow.go (hunk 15 above)
Comment (mickmister) [reply server/setup_flow.go:648]: There's not really a "new" instance here. There will always be an existing one in the kv store at this point of the setup. There are multiple stages to the setup flow and this is the second one
